wiki/multilang_md.py

```diff
@@ -0,0 +1,333 @@
+
+    def write(self, oneline, codeblock_mark):
+        # write
+        self.content.append(oneline)
+        # check header
+        if not codeblock_mark:
+            result = self.header_re.match(oneline)
+            if result:
+                head = oneline[result.end():].replace("\n", "")
+                self.append_toc(head, result.end() - 1)
+
+    def append_toc(self, head, level):
+        if (self.prev_level + 1) == level:  # new child
+            pass
+        elif self.prev_level == level:  # same level
+            self.toc_stack = self.toc_stack[:-1]
+        elif self.prev_level > level:  # new elem
+            self.toc_stack = self.toc_stack[: (level - 1)]
+        else:
+            msg = "'Table of Contents' level mismatch. prev_level="
+            msg += f"{self.prev_level}, current_level={level}"
+            raise self.CustomException(msg)
+        recursive = self.toc
+        for x in self.toc_stack:
+            recursive = recursive[x]
+        recursive[head] = OrderedDict()
+        self.toc_stack.append(head)
+        self.prev_level = level
+
+    def set_toc(self, _min=1, _max=9, enable_emoji_header=True):
+        self.content.append((_min, _max, enable_emoji_header))
+
+
+class MultilangualDoc(object):
+    class SafetyRead(object):
+        def __init__(self, full_name):
+            self.__doc = open(full_name, "r", encoding="utf-8")
+            self.line_count = 0
+            self.last_line = ""
+
+        def __del__(self):
+            self.__doc.close()
+
+        def readline(self):
+            self.line_count += 1
+            self.last_line = self.__doc.readline()
+            return self.last_line
+
+    def __init__(self, path, file_name):
+        self._base_file = file_name
+        self.full_name = os.path.join(path, self._base_file)
+        self.doc = self.SafetyRead(self.full_name)
+
+        # Initialize configuration
+        self.lang_read = {}
+        self.lang_doc = {}
+        self.no_suffix = ""
+        self.pattern = re.compile(r"[ \t]*<!-- \[\w+\] -->")  # <!-- [oo] -->
+        self.init_config(path, self._base_file)
+
+        # Main converting
+        self.converting()
+        self.log()
+
+    def CustomException(self, message):
+        return click.ClickException(f"<{self._base_file}> {message}")
+
+    def log(self):
+        click.echo(self.full_name)
+        for lang, md in self.lang_doc.items():
+            click.echo(f"\t{lang}: {md}")
+
+    def init_config(self, path, file_name):
+        # re options
+        options = [
+            (
+                re.compile(r"<!-- multilangual suffix:[ ]*([\w]+)(, [\w]+)*"),
+                self.config_setlang,
+            ),
+            (re.compile(r"<!-- no suffix:[ ]*([\w]+)"), self.config_nosuffix),
+        ]
+        option_checked = [False for _ in options]
+        # check base doc
+        while True:
+            for i in range(len(options)):
```

Comment:
Codacy found an issue: [Consider using enumerate instead of iterating with range and len](https://app.codacy.com/gh/mommel/hs-lr-midi-schwurbler/pullRequest?prid=6419091)